### PR TITLE
Changes .length & .count to .size

### LINKED ISSUE #8205: Replace .count and .length in tag model with .size
Hi, this is a [first-timers-only issue](https://code.publiclab.org/#r=all). This means we've worked to make it more legible to folks who either **haven't contributed to our codebase before, or even folks who haven't contributed to open source before**. 

If that's you, we're interested in helping you take the first step and can answer questions and help you out as you do. Note that we're especially interested in contributions from people from groups underrepresented in free and open source software!

We know that the process of creating a pull request is the biggest barrier for new contributors. This issue is for you 💝

If you have contributed before, **consider leaving this one for someone new**, and looking through our general [help wanted](https://github.com/publiclab/plots2/labels/help-wanted) issues. Thanks!

### 🤔 What you will need to know.

Nothing. This issue is meant to welcome you to Open Source :) We are happy to walk you through the process.

### 📋 Step by Step

- [ ] 🙋 **Claim this issue**: Comment below. If someone else has claimed it, ask if they've opened a pull request already and if they're stuck -- maybe you can help them solve a problem or move it along!

- [ ] 📝 **Update** the file  https://github.com/publiclab/plots2/blob/main/app/models/tag.rb   (press the little pen Icon) and edit the line as shown below. 
-  Replace all instances of `.count` and `.length` with `.size`

[See this page](https://code.publiclab.org/#r=all) for some help in taking your first steps!

Below is a "diff" showing in red (and a `-`) which lines to remove, and in green (and a `+`) which lines to add:

```diff
$DIFF
```

- [ ] 💾 **Commit** your changes

- [ ] 🔀 **Start a Pull Request**. There are two ways how you can start a pull request:

1. If you are familiar with the terminal or would like to learn it, [here is a great tutorial](https://egghead.io/series/how-to-contribute-to-an-open-source-project-on-github) on how to send a pull request using the terminal.

2. You can also [edit files directly in your browser](https://help.github.com/articles/editing-files-in-your-repository/) and open a pull request from there. 

- [ ] 🏁 **Done** Ask in comments for a review :)

### Please keep us updated

💬⏰ - We encourage contributors to be respectful to the community and provide an update **within a week** of claiming a first-timers-only issue. We're happy to keep it assigned to you as long as you need if you update us with a request for more time or help, but if we don't see any activity a week after you claim it we may reassign it to give someone else a chance. Thank you in advance! 

If this happens to you, don't sweat it! Grab another open issue.

### Is someone else already working on this?

🔗- We encourage contributors to link to the original issue in their pull request so all users can easily see if someone's already started on it. 

👥- **If someone seems stuck, offer them some help!** Otherwise, [take a look at some other issues you can help with](https://code.publiclab.org/#r=all). Thanks!

### 🤔❓ Questions?

Leave a comment below!

### PR DESCRIPTION
**Pull request by Hassanelashram**
I changed all the instances as asked.
At first I misunderstood and created a branch locally where I edited all the .count and .length in ruby files in the whole codebase.
so if that is something that needs to be done at some point let me know and I could make a PR with the changes as well.

Fixes #8205  (<=== Add issue number here)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [x] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

**Comment by avats-dev** [comment]
@Hassanelashram thanks for working on this. Could you mention the concerned issue in your PR, as it will help others review. Thanks :+1:

**Comment by Hassanelashram** [comment]
> @Hassanelashram thanks for working on this. Could you mention the concerned issue in your PR, as it will help others review. Thanks 👍

Hey, yes sorry, it's my first time contributing in an open source: https://github.com/publiclab/plots2/pull/8402#issuecomment-694248678

**Comment by avats-dev** [comment]
> > @Hassanelashram thanks for working on this. Could you mention the concerned issue in your PR, as it will help others review. Thanks +1
> 
> Hey, yes sorry, it's my first time contributing in an open source: [#8402 (comment)](https://github.com/publiclab/plots2/pull/8402#issuecomment-694248678)

Hey @Hassanelashram , thanks for your contribution and welcome to publiclab community. We want to help first timers contribute. You can always ping back if you're stuck or need any help.

#8402 is this PR, not the issue you're resolving. See this:
> It's very important to link your PR to the issue that it intends to resolve, as this can:
> 
>     * help others check if PR is any good for what it is intended.
> 
>     * automatically close that issue when the PR is merged.
> 
>     * help in better workflow of the codebase.
> 
> 
> To link your PR to the issue, edit your comment (the first one, i.e. one with which you have opened this PR) and you'll see something like "**Fixes #0000 (<=== Add issue number here)**". This is even the first line of your comment. You have to replace **0000** here with the issue number concerned with your PR, and leave everything else as it is. **Don't remove the # as it is crucial and creates the link in between.**
> 
> If you're still unsure and having problems, watch this as an example.
> Peek 2020-08-22 23-58

File: app/models/tag.rb [hunk 1]
```diff
@@ -443,7 +445,7 @@ def nids
   end
 
   def span(start, fin)
-    start.to_date.step(fin.to_date, 7).count
+    start.to_date.step(fin.to_date, 7).size
```
Comment (cesswairimu) [inline app/models/tag.rb:448]: hello, seems a couple of things won't work here. Lets leave this as `.count`  `.size` will return nil here

File: app/models/tag.rb [hunk 2]
```diff
@@ -412,7 +421,7 @@ def self.all_tags_by_popularity
 
   def subscription_graph(start = DateTime.now - 1.year, fin = DateTime.now)
     date_hash = {}
-    week = start.to_date.step(fin.to_date, 7).count
+    week = start.to_date.step(fin.to_date, 7).size
```
Comment (cesswairimu) [inline app/models/tag.rb:424]: lets also leave `.count` the same here

File: app/models/tag.rb [hunk 3]
```diff
@@ -359,18 +367,19 @@ def self.tagged_node_count(tag_name)
...
-                    .limit(5)
-                    .pluck(:nid)
+      nids = NodeTag.joins(:tag)
+                     .where(Tag.table_name => { name: tag_name })
+                     .select(:nid)
+
+      # sort them by how often they co-occur:
+      nids = nids.group_by{ |v| v }.map{ |k, v| [k, v.size] }
+      nids = nids.collect(&:first)[0..4]
+                 .collect(&:nid) # take top 5
```
Comment (cesswairimu) [inline app/models/tag.rb:382]: Not sure if you updated this by mistake. Lets please leave it as is

File: app/models/tag.rb [hunk 4]
```diff
@@ -261,7 +269,7 @@ def comment_graph(start = Time.now - 1.year, fin = Time.now)
     while week >= 1
       month = (fin - (week * 7 - 1).days)
       weekly_comments = comments.where(timestamp: range(fin, week))
-        .count(:all)
+        .size(:all)
```
Comment (cesswairimu) [inline app/models/tag.rb:272]: lets also leave this as `.count`... `.size(:all)` is not a valid rails method

File: app/models/tag.rb [hunk 5]
```diff
@@ -245,9 +253,9 @@ def quiz_graph(start = Time.now - 1.year, fin = Time.now)
     while week >= 1
       month = (fin - (week * 7 - 1).days)
       weekly_quiz = questions.where(created: range(fin, week))
-        .count(:all)
+        .size(:all)
```
Comment (cesswairimu) [inline app/models/tag.rb:256]: lets also leave this as .count... .size(:all) is not a valid rails method

File: app/models/tag.rb [hunk 6]
```diff
@@ -229,7 +237,7 @@ def contribution_graph_making(type = 'note', start = Time.now - 1.year, fin = Ti
           nids,
           (fin.to_i - week.weeks.to_i).to_s,
           (fin.to_i - (week - 1).weeks.to_i).to_s
-        ).count(:all)
+        ).size(:all)
```
Comment (cesswairimu) [inline app/models/tag.rb:240]: lets also leave this as .count... .size(:all) is not a valid rails method

File: app/models/tag.rb [hunk 7]
```diff
@@ -207,7 +215,7 @@ def weekly_tallies(type = 'note', span = 52)
         nids,
         (Time.now.to_i - week.weeks.to_i).to_s,
         (Time.now.to_i - (week - 1).weeks.to_i).to_s
-      ).count(:all)
+      ).size(:all)
```
Comment (cesswairimu) [inline app/models/tag.rb:218]: lets also leave this as .count... .size(:all) is not a valid rails method

File: app/models/tag.rb [hunk 8]
```diff
@@ -37,7 +37,7 @@ def path
   end
 
   def run_count
-    self.count = NodeTag.joins(:node).where(tid: tid).where('node.status = 1').count
+    self.size = NodeTag.where(tid: tid).size
```
Comment (cesswairimu) [inline app/models/tag.rb:40]: lets leave the check for status here

` self.count = NodeTag.joins(:node).where(tid: tid).where('node.status = 1').size`

**Review by cesswairimu — CHANGES_REQUESTED** [review]
Sorry forgot to indicate the appropriate places that the replacement will work.

**Comment by Hassanelashram** [comment]
@cesswairimu Thank you so much! I fixed them now, regarding the checks, it seems like rubocop is failing for a missing space ?

**Comment by cesswairimu** [comment]
Thanks. Not sure why travis is still failing, I will pull this and check locally if there are other replacements that are not working. 
As for rubocop yes it recommends we have space before and after `{` and `}`

**Comment by Hassanelashram** [comment]
> Thanks. Not sure why travis is still failing, I will pull this and check locally if there are other replacements that are not working.
> As for rubocop yes it recommends we have space before and after `{` and `}`

Cool thanks, 
and for the spaces I am confused as I see that line 380 already has the spaces @cesswairimu

File: app/models/tag.rb [hunk 9]
```diff
@@ -364,13 +372,14 @@ def self.tagged_node_count(tag_name)
...
-                    .group(:nid)
-                    .order(NodeTag.arel_table[:nid].count.desc)
-                    .limit(5)
-                    .pluck(:nid)
+      nids = NodeTag.joins(:tag)
+                     .where(Tag.table_name => { name: tag_name })
+                     .select(:nid)
+
+      # sort them by how often they co-occur:
+      nids = nids.group_by{ |v| v }.map{ |k, v| [k, v.size] }
```
Comment (cesswairimu) [inline app/models/tag.rb:380]: No worries @Hassanelashram , what I meant is this should be like 
`nids = nids.group_by { |v| v }.map { |k, v| [k, v.size] }`

**Comment by cesswairimu** [comment]
Hi @Hassanelashram, we encourage contributors to take one first-timer-only issue( These are issues with that label) and leave the others for new contributors coming in. Since this PR we merger one, closing this to give another new contributor a chance to work on it. You can look for issues with the label 'help-wanted' .Thanks

**Pull request closed without merging**